lib/AST/ASTScopeCreation.cpp

```diff
@@ -0,0 +1,1061 @@
+//===--- ASTScopeCreation.cpp - Swift Object-Oriented AST Scope -----------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2017 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+///
+/// This file implements the creation methods of the ASTScopeImpl ontology.
+///
+//===----------------------------------------------------------------------===//
+#include "swift/AST/ASTScope.h"
+
+#include "swift/AST/ASTContext.h"
+#include "swift/AST/ASTWalker.h"
+#include "swift/AST/Decl.h"
+#include "swift/AST/Expr.h"
+#include "swift/AST/Initializer.h"
+#include "swift/AST/LazyResolver.h"
+#include "swift/AST/Module.h"
+#include "swift/AST/ParameterList.h"
+#include "swift/AST/Pattern.h"
+#include "swift/AST/Stmt.h"
+#include "swift/AST/TypeRepr.h"
+#include "swift/Basic/STLExtras.h"
+#include "llvm/Support/Compiler.h"
+#include <algorithm>
+
+using namespace swift;
+namespace swift {
+namespace ast_scope {
+
+#pragma mark DeferredNodes
+
+/// \c SourceFiles and \c BraceStmts have \c Decls and \c ASTNodes in them that
+/// must be scoped in distant descendants.
+/// As scopes are created, pass down the \c ASTNodes in enclosing \c Decls
+/// that really need to be in subscopes.
+///
+/// Because parsing and lookup (typechecking) are interleaved for some files,
+/// we need to be able to resume scope tree creation. \c DeferredNodes also
+/// helps with this by remembering the \c lastAdopter.
+
+class DeferredNodes final {
+  /// The nodes that must be passed down to deeper levels in the scope tree.
+  std::deque<ASTNode> nodes;
+
+  /// The number of \c Decls in the \c SourceFile that were already processed
+  /// into scopes. Since parsing can be interleaved with type-checking, on every
+  /// lookup, create scopes for any \c Decls beyond this number.
+  int numberOfDeclsAlreadyProcessed = 0;
+
+  /// The last scope to "adopt" deferred nodes.
+  /// When adding \c Decls to a scope tree that have been created since the tree
+  /// was originally built, add them as children of this scope.
+  NullablePtr<ASTScopeImpl> lastAdopter;
+
+  /// Information about how this object was created for debugging.
+  const char *const file;
```

Comment:
If this is for debugging, it should be ifdefed out without asserts. Also, you probably want to use a StringRef.

Reply:
Yes, agree about the ifdeffing. What's the benefit of a `StringRef` over a `const char*` here? These are truly compile-time constants, being the filename of the .cpp file in the compiler. A `const char*` is cheaper here, not that it matters much. 
I'll add some comments to make it clear that `file` is the compiler file.